### PR TITLE
bpo-29762: More use "raise from None".

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
This hides unwanted implementation details from tracebacks.

### CODE REVIEW

File: Lib/encodings/punycode.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ def decode_generalized_number(extended, extpos, bias, errors):
             char = ord(extended[extpos])
         except IndexError:
             if errors == "strict":
-                raise UnicodeError("incomplete punicode string")
+                raise UnicodeError("incomplete punicode string") from None
```
Comment (wm75) [inline Lib/encodings/punycode.py]: information about extpos is lost here

File: Lib/mailbox.py [hunk 2]
```diff
@@ -201,7 +201,7 @@ def _string_to_bytes(self, message):
             return message.encode('ascii')
         except UnicodeError:
             raise ValueError("String input must be ASCII-only; "
-                "use bytes or a Message instead")
+                "use bytes or a Message instead") from None
```
Comment (wm75) [inline Lib/mailbox.py]: information about problematic character in the string is lost

**Review by wm75 — CHANGES_REQUESTED** [review]
Generally, I think this is a good idea, but only if there is absolutely no loss of information, which happens at least for two of your changes (see comments, there may be others that I've missed)

File: Lib/encodings/punycode.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/encodings/punycode.py]: Done. And there are other bugs in this codec.

File: Lib/mailbox.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/mailbox.py]: Done.

**Comment by vstinner** [comment]
FYI It's hard to review such change, because each change should be carefully reviewed, whereas the change modifies 31 files. So I decided to skip my turn, maybe review it another day :-)

File: Lib/json/tool.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def main():
                 obj = json.load(infile,
                                 object_pairs_hook=collections.OrderedDict)
         except ValueError as e:
-            raise SystemExit(e)
+            raise SystemExit(e) from None
```
Comment (rhettinger) [inline Lib/json/tool.py]: I'm note sure about this one.

**Comment by rhettinger** [comment]
Except for the one noted in the comment, these all look good.

File: Lib/json/tool.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/json/tool.py]: Agree. It just doesn't have any visible effect. The original exception is passes to SystemExit, and `str(e)` is printed to stderr at exit. `__cause__` and `__context__` are ignored.

**Comment by serhiy-storchaka** [comment]
Thank you for your review @wm75 and @rhettinger.

**Pull request merged**